golang/citadel.go

```diff
@@ -0,0 +1,153 @@
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"io/ioutil"
+	"encoding/json"
+	"os"
+	"bytes"
+)
+
+type PublicTokenRequest struct {
+	PublicTokens []string `json:"public_tokens"`
+}
+
+type AccessTokenRequest struct {
+	AccessToken string `json:"access_token"`
+}
+
+type AccessTokenResponse struct {
+	AccessTokens []string `json:"access_tokens"`
+}
+
+func getRequest(endpoint string, method string, body []byte) (*http.Request) {
+	apiUrl := os.Getenv("API_URL")
+	clientId := os.Getenv("API_CLIENT_ID")
+	accessKey := os.Getenv("API_SECRET")
+	fullEndpoint := fmt.Sprintf("%s%s", apiUrl, endpoint)
+	request, _ := http.NewRequest(method, fullEndpoint, bytes.NewBuffer(body))
+	request.Header.Set("Content-Type", "application/json")
+	request.Header.Set("X-Access-Client-Id", clientId)
+	request.Header.Set("X-Access-Secret", accessKey)
+	return request
+}
+
+func getBridgeToken() (string) {
+	request := getRequest("bridge-tokens/", "POST", nil)
+	client := &http.Client{}
+	response, err := client.Do(request)
+	if err != nil {
+    fmt.Printf("The HTTP request failed with error %s\n", err)
+	} else {
+		data, _ := ioutil.ReadAll(response.Body)
+		return (string(data))
+	}
+	return ""
+}
+
+func getAccessToken(public_token string) (string) {
+	publicTokens := PublicTokenRequest{PublicTokens: []string{public_token}}
+	jsonPublicTokens, _ := json.Marshal(publicTokens)
+	accessTokens := AccessTokenResponse{}
+	request := getRequest("access-tokens/", "POST", jsonPublicTokens)
+	client := &http.Client{}
+	res, err := client.Do(request)
+    defer res.Body.Close()
+
+    if err != nil {
+        fmt.Printf("The HTTP request failed with error %s\n", err)
+		}
+    err = json.NewDecoder(res.Body).Decode(&accessTokens)
+    if err != nil {
+        panic(err)
+		}
+    return accessTokens.AccessTokens[0]
+}
+
+func getEmploymentInfoByToken(access_token string) (string) {
+	accessToken := AccessTokenRequest{AccessToken: access_token}
+	jsonAccessToken, _ := json.Marshal(accessToken)
+	request := getRequest("verifications/employments", "POST", jsonAccessToken)
+	client := &http.Client{}
+	res, err := client.Do(request)
+    defer res.Body.Close()
+
+    if err != nil {
+        fmt.Printf("The HTTP request failed with error %s\n", err)
+		}
+    data, _ := ioutil.ReadAll(res.Body)
+    return string(data)
+}
+
+func getIncomeInfoByToken(access_token string) (string) {
+	accessToken := AccessTokenRequest{AccessToken: access_token}
+	jsonAccessToken, _ := json.Marshal(accessToken)
+	request := getRequest("verifications/incomes", "POST", jsonAccessToken)
+	client := &http.Client{}
+	res, err := client.Do(request)
+    defer res.Body.Close()
+
+    if err != nil {
+        fmt.Printf("The HTTP request failed with error %s\n", err)
```

Comment:
in go every function should return error + something. e.g. here
```
func getIncomeInfoByToken(access_token string) (*string, error) {
```

in case of success:
```
return string(data), nil
```

in case of error:
```
return nil, error 
```

but now your code will fail in case of http errors (it will print smth here, and then will try to return body)